chainerx_cc/chainerx/routines/binary.cc

```diff
@@ -0,0 +1,125 @@
+#include "chainerx/routines/binary.h"
+
+#include "chainerx/array.h"
+#include "chainerx/backprop_mode.h"
+#include "chainerx/backward_builder.h"
+#include "chainerx/backward_context.h"
+#include "chainerx/dtype.h"
+#include "chainerx/error.h"
+#include "chainerx/graph.h"
+#include "chainerx/kernels/binary.h"
+#include "chainerx/routines/routines_util.h"
+#include "chainerx/routines/type_util.h"
+#include "chainerx/scalar.h"
+#include "chainerx/shape.h"
+
+namespace chainerx {
+namespace {
+
+void CheckBitwiseDtypes(DtypeKind kind1, DtypeKind kind2) {
+    if (kind1 == DtypeKind::kFloat || kind2 == DtypeKind::kFloat) {
+        throw DtypeError{"Bitwise operations don't support Float types"};
+    }
+}
+
+void CheckInplaceBitwiseDtypes(DtypeKind out_kind, DtypeKind in_kind) {
+    CheckBitwiseDtypes(out_kind, in_kind);
+    if (out_kind == DtypeKind::kBool && in_kind != DtypeKind::kBool) {
+        throw DtypeError{"Bitwise operations don't allow updating Bool array with integer array"};
+    }
+}
+
+void CheckBitwiseDtypes(const Array& x1, const Array& x2) { CheckBitwiseDtypes(GetKind(x1.dtype()), GetKind(x2.dtype())); }
+
+void CheckBitwiseDtypes(const Array& x1, const Scalar& x2) { CheckBitwiseDtypes(GetKind(x1.dtype()), x2.kind()); }
+
+void CheckInplaceBitwiseDtypes(const Array& x1, const Array& x2) { CheckInplaceBitwiseDtypes(GetKind(x1.dtype()), GetKind(x2.dtype())); }
+
+void CheckInplaceBitwiseDtypes(const Array& x1, const Scalar& x2) { CheckInplaceBitwiseDtypes(GetKind(x1.dtype()), x2.kind()); }
+
+template <typename Impl>
+void BitwiseImpl(const Array& x1, const Array& x2, const Array& out) {
+    CheckEqual(x1.shape(), x2.shape());
```

Comment:
This should be an assertion?

Reply:
Yes. `internal::BroadcastBinary` broadcasts the inputs beforehand.